saml/src/main/java/com/linecorp/armeria/server/saml/SamlService.java

```diff
@@ -0,0 +1,171 @@
+                                                                sp.sloHandler())));
+        final PathMapping metadata = sp.metadataPath();
+        metadata.exactPath().ifPresent(
+                path -> builder.put(path,
+                                    new SamlMetadataServiceFunction(sp.entityId(),
+                                                                    sp.signingCredential(),
+                                                                    sp.encryptionCredential(),
+                                                                    sp.idpConfigs(),
+                                                                    sp.acsConfigs(),
+                                                                    sp.sloEndpoints())));
+        serviceMap = builder.build();
+        pathMappings = serviceMap.keySet().stream().map(PathMapping::ofExact).collect(toImmutableSet());
+    }
+
+    @Override
+    public void serviceAdded(ServiceConfig cfg) throws Exception {
+        if (server != null) {
+            if (server != cfg.server()) {
+                throw new IllegalStateException("cannot be added to more than one server");
+            } else {
+                return;
+            }
+        }
+
+        server = cfg.server();
+
+        // Auto-detect the primary port number and its session protocol after the server started.
+        server.addListener(portConfigHolder);
+    }
+
+    @Override
+    public Set<PathMapping> pathMappings() {
+        return pathMappings;
+    }
+
+    @Override
+    public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        final SamlServiceFunction func = serviceMap.get(req.path());
+        if (func == null) {
+            return HttpResponse.of(HttpStatus.BAD_REQUEST);
+        }
+
+        final CompletionStage<AggregatedHttpMessage> f;
+        if (portConfigHolder.isDone()) {
+            f = req.aggregate();
+        } else {
+            f = portConfigHolder.future().thenCompose(unused -> req.aggregate());
+        }
+
+        // SAML processing will be done over the blocking task executor because there are lots of
+        // XML processing.
+        return HttpResponse.from(f.handleAsync((msg, cause) -> {
+            if (cause != null) {
+                return HttpResponse.of(HttpStatus.BAD_REQUEST);
+            }
+            final SamlPortConfig portConfig = portConfigHolder.config().get();
+            if (portConfig.scheme().isTls() != ctx.sessionProtocol().isTls()) {
+                return HttpResponse.of(HttpStatus.BAD_REQUEST);
+            }
+
+            // Use user-specified hostname if it exists.
+            // If there's no hostname set by a user, the default virtual hostname will be used.
+            final String defaultHostname = firstNonNull(sp.hostname(), ctx.virtualHost().defaultHostname());
+            return func.serve(ctx, msg, defaultHostname, portConfig);
+        }, ctx.blockingTaskExecutor()));
```

Comment:
Sorry for missing comments. I also wanted to prevent event loops from a user's blocking calls in the user handlers, such as `SamlSingleSignOnHandler` and `SamlSingleLogoutHandler`. 

I also talked with @trustin about this, then we make every callback methods return asynchronous like `CompletionStage<Void>` instead of `void`, in order to lower the possibility of blocking event loop.

Thanks.

Reply:
Oops! I missed fixing here. 😱 Anyway, the reason why I used `blockingTaskExecutor()` here was due to worry about slow XML handling and user's blocking calls in user handlers. Will replace `handleAsync` with `handle`. Thanks!